agent/util-scripts/pbench-tool-meister-start

```diff
@@ -527,150 +616,333 @@ def main(argv):
                 failures += 1
-                successes -= 1
-                logger.error(
-                    "INTERNAL ERROR: os.waitpid(%d, 0) returned (%d, %d [%0X])",
-                    pid,
-                    exit_pid,
-                    exit_status,
-                    _exit_status,
+                continue
+            pid = tm_proc["pid"]
+            try:
+                exit_status = waitpid(pid)
+            except Exception:
+                failures += 1
+                logger.exception(
+                    "failed to create a tool meister instance for host %s", host
                 )
             else:
                 if exit_status != 0:
                     failures += 1
-                    successes -= 1
                     logger.error(
                         "failed to start tool meister on remote host '%s'"
-                        " (pid %d), exit status: %d [%0X]",
+                        " (pid %d), exit status: %d",
                         host,
                         pid,
                         exit_status,
-                        _exit_status,
                     )
+                else:
+                    successes += 1
 
-    if failures > 0:
-        logger.info("terminating tool meister startup due to failures")
-        terminate_msg = dict(action="terminate", group=group, directory=None)
-        try:
-            ret = redis_server.publish(
-                channel, json.dumps(terminate_msg, sort_keys=True)
+        if failures > 0:
+            # Don't wait for the Tool Meisters
+            logger.info("terminating tool meister startup due to failures")
+            terminate_msg = dict(action="terminate", group=group, directory=None)
+            try:
+                ret = redis_server.publish(
+                    f"{cli_tm_channel_prefix}-{tm_channel_suffix_from_client}",
+                    json.dumps(terminate_msg, sort_keys=True),
+                )
+            except Exception:
+                logger.exception("Failed to publish terminate message")
+            else:
+                logger.debug("publish('terminate') = %r", ret)
+            return kill_redis_server(redis_pid, ReturnCode.TMFAILURES)
+
+        if successes == 0:
+            logger.warning(
+                "unable to successfully start any tool meisters,"
+                " but encountered no failures either: terminating"
             )
-        except Exception:
-            logger.exception("Failed to publish terminate message")
-        else:
-            logger.debug("publish() = %r", ret)
-        ret_val = kill_redis_server(redis_pid)
-    elif successes > 0:
-        # If any successes, then we need to wait for them to show up as
-        # subscribers.
-        logger.debug(
-            "e. waiting for all successfully spawned SSH processes"
-            " to show up as subscribers"
+            return kill_redis_server(redis_pid, ReturnCode.TMNOSUCCESSES)
```

Comment:
Consider having one place to kill the Redis server and then return the intended code perhaps stored in the exception object.

Reply:
Tracked in https://github.com/distributed-system-analysis/pbench/issues/2166.